extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/BigQueryServiceImpl.java

```diff
@@ -0,0 +1,594 @@
+            ((OAuth2Credentials) credentials).refreshIfExpired();
+        } catch (IOException ioException) {
+            monitor.warning("BigQuery Service cannot refresh credentials", ioException);
+        }
+    }
+
+    private void outputPage(OutputStream outputStream, TableResult tableResult) throws IOException {
+        var page = new JSONArray();
+
+        tableResult.getValues()
+                .forEach(row -> page.put(buildRecord(tableResult.getSchema().getFields(), row)));
+
+        outputStream.write(page.toString().getBytes());
+    }
+
+    private void setNamedParameters(QueryJobConfiguration.Builder queryConfigBuilder, DataAddress sinkAddress) {
+        var prefix = EDC_NAMESPACE + '@';
+        for (var key : sinkAddress.getProperties().keySet()) {
+            if (key.startsWith(prefix)) {
+                key = key.substring(EDC_NAMESPACE.length());
+
+                var value = sinkAddress.getStringProperty(key);
+                key = key.substring(1);
+                int separatorIndex = key.indexOf("_");
+                if (separatorIndex != -1) {
+                    var type = key.substring(0, separatorIndex);
+                    key = key.substring(separatorIndex + 1);
+                    var sqlValue = QueryParameterValue.of(value, StandardSQLTypeName.valueOf(type));
+                    queryConfigBuilder.addNamedParameter(key, sqlValue);
+                }
+            }
+        }
+    }
+
+    @Override
+    public Stream<DataSource.Part> runSourceQuery(String query, DataAddress sinkAddress) throws InterruptedException {
+        refreshBigQueryCredentials();
+
+        var queryConfigBuilder = QueryJobConfiguration.newBuilder(query);
+        var customerName = sinkAddress.getStringProperty(CUSTOMER_NAME);
+        if (customerName != null) {
+            queryConfigBuilder.setLabels(ImmutableMap.of("customer", customerName));
+        }
+
+        setNamedParameters(queryConfigBuilder, sinkAddress);
+
+        // Use standard SQL syntax for queries.
+        // See: https://cloud.google.com/bigquery/sql-reference/
+        var queryConfig = queryConfigBuilder
+                .setUseLegacySql(false)
+                .setUseQueryCache(true)
+                .build();
+
+        var jobId = JobId.of(UUID.randomUUID().toString());
+        var jobInfo = JobInfo.newBuilder(queryConfig).setJobId(jobId).build();
+        final var queryJob = bigQuery.create(jobInfo);
+
+        var parts = new ArrayList<DataSource.Part>();
+        // The output stream object is passed to the thread lambda, do not use try with resources.
+        final var outputStream = new PipedOutputStream();
+        try {
+            // The input stream object is passed to the sink thread, do not use try with resources.
+            var inputStream = new PipedInputStream(outputStream);
+            parts.add(new BigQueryPart("allRows", inputStream));
+
+            new Thread(() -> {
```

Comment:
please use an executor service instead of instantiating the `Thread` directly so it will be possible to keep track of the pool and to configure the size and so on.

Reply:
Done